python/lsst/daf/butler/_butler.py

```diff
@@ -855,7 +857,32 @@ def put(self, obj: Any, datasetRefOrType: Union[DatasetRef, DatasetType, str],
 
         # Add Registry Dataset entry.
         dataId = self.registry.expandDataId(dataId, graph=datasetType.dimensions, **kwds)
-        ref, = self.registry.insertDatasets(datasetType, run=run, dataIds=[dataId])
+
+        # For an execution butler the datasets will be pre-defined.
+        # If the butler is configured that way datasets should only be inserted
+        # if they do not already exist in registry. Trying and catching
+        # ConflictingDefinitionError will not work because the transaction
+        # will be corrupted. Instead, in this mode always check first.
+        ref = None
+        if self._allow_put_of_predefined_dataset:
+            # Get the matching ref for this run.
+            existing_refs = list(self.registry.queryDatasets(datasetType, collections=run,
+                                                             dataId=dataId))
```

Comment:
I'd use `Registry.findDataset` instead here; that will be more efficient, and it will always return either a resolved ref or `None`.  Getting more than that should be impossible anyway due to database unique constraints.

I also think it'd be a lot cleaner if we actually put the resolved output refs in the QuantumGraph and only enabled this logic if the ref was already resolved (or even just called `butler.datastore.put`).  But that involves making the execution butler before we actually write out the QG (or generating UUIDs for predicted datasets out the butler, once we can assume UUIDs everwhere), so it's probably not worth worrying about right now.

Reply:
Yes, Sorry. Will switch. I remembered getDataset but that needed the ID.  

I also just had to push a change that is bigger than I expected because writing the quick test demonstrated that calling datastore.put leads to the data file being deleted if it goes wrong and that is very bad if you were relying on the file already being there (and it also overwrites it). I've added a `datastore.knows()` API to work around that.

Reply:
New changes look good, too, though all of this special handling for execution butlers reinforces my thinking that we want to do something (like QuantumDirectory) to let us drop it someday.  But not today.  Merge away, whenever you're ready.